Feat: call worker on receive report

moreirathomas opened the pull request:
<!-- IMPORTANT: Don't forget to link the issue(s) once the PR created! -->

## Description

Save in Firestore reports posted by the `runner`. Use cloud.google.com/go/firestore library.

Google Cloud uses [service accounts](https://cloud.google.com/iam/docs/service-accounts#key-expiry) to authenticate an application requesting access to a cloud ressource. Here we ask for read and write access to Firestore.

In a fully deployed environment, the server application would run inside Cloud Run (Cloud Engine? _todo: find the correct one for our needs_) and Google Cloud would under the hood find and use a default service account correctly authenticated (see [Finding credentials automatically](https://cloud.google.com/docs/authentication/production#automatically)).

However in dev mode, the app runs locally and we must provide a service account credentials manually inside the `GOOGLE_APPLICATION_CREDENTIALS` env variable (internally read by `firestore.NewClient`). To do that, follow the steps from [Passing credentials manually](https://cloud.google.com/docs/authentication/production#manually).

The service account will need read and write access to Firestore, which can be achieve by giving it the role `roles/datastore.user` as suggested [here](https://cloud.google.com/firestore/docs/security/iam). 

Make sure you read the security related docs!


## Changes

- the domain exposes `PersitenceAdapter`, an interface defining how to interact with the persistence layer (for now `Report` in Firestore, later `Stats` in SQL Cloud)
- keep previous `repository` as `ondisk` adapter, `store` is renamed `kvstore` (the feature may be drop before merging, although it may prove it's use and works fine)
- flatten the file structure
- update `Report` model to match `benchttp/runner` and copy-paste all required structs (all structs children of `benchttp/runner/output.Output`)
- use `gob` to decode data sent by `benchttp/runner`

## Notes

With the flattening of the file structure, `Report` is define at the root inside package `server`, but `server.Report` is a super weird name. We might want to think about that (some day?).

Also, there's a lot of copy-paste regarding struct definitions, which is _meh_.
By using `gob` between `runner` and `server`, the structs are _coupled across repos_, e.g. the fact that we need to change `runner` field types because Firestore does not support it (benchttp/runner#69). But `runner` should not really know, care or adapt to that.
Subject to discussion.
 
~🚧 not ready for merge, as `worker` currently only accepts `Report` from Firestore, thus `server` uploads `Report` and not `Output`, yet.~

<!-- Optional: additional notes than can help understanding the implementation -->
_May_ close #10

GregoryAlbouy commented on .env.example:
```diff
@@ -0,0 +1,2 @@
+PROJECT_ID=your-project-id
```
`GOOGLE_PROJECT_ID` / `GCP_PROJECT_ID`?

GregoryAlbouy commented on .env.example:
```diff
@@ -0,0 +1,2 @@
+PROJECT_ID=your-project-id
+GOOGLE_APPLICATION_CREDENTIALS=pathto/service-account-file.json
```
```suggestion
GOOGLE_APPLICATION_CREDENTIALS=pathto/service-account-file.json

```

GregoryAlbouy commented on .gitignore:
```diff
@@ -4,3 +4,4 @@
 # Env files
 .env
 .data/
+.*keys*
```
[question] What is this file?

✨  Bonus suggestion ✨ 
```suggestion
.*keys*

```

GregoryAlbouy commented on adapter.go:
```diff
@@ -0,0 +1,14 @@
+
+// PersistenceAdapter defines the interface to implement by a
+// data layer adapter inside the application.
+type PersistenceAdapter interface {
```
[nitpick] Nice name, but maybe a bit misleading because there's no adaptation involved in the described methods (it's only about storing and retrieving a report, what I understand here is the "adapter" part is just the fact it's an interface). Why not the classic `ReportService`?

Edit:
> - the domain exposes `PersistenceAdapter`, an interface defining how to interact with the persistence layer (for now `Report` in Firestore, later `Stats` in SQL Cloud)

So you're considering using that single interface for two distinct services? How can we do that without losing the typing? 🤔  Also forcing them to share the same methods seems like a huge constraint -- I feel I'm totally missing something here, need some details 😅

moreirathomas replied:
As for now there's only one data model store I grouped everything in the same interface, but I'm with you the implementation details would define another interface when there are multiple data structures.

My wording was not clear ^^

That or the method are suffixed `Persistence.CreateFoo` and `Persistence.CreateBar` which is ok but not ideal.

GregoryAlbouy commented on cmd/main.go:
```diff
@@ -15,7 +18,17 @@ func main() {
-	repo, err := repository.New()
+	err := godotenv.Load(".env")
+	if err != nil {
+		log.Fatalf("Error loading .env file")
```
[yolo] Surprised the linter doesn't yell!

GregoryAlbouy commented on firestore/repository.go:
```diff
@@ -0,0 +1,55 @@
+func NewRepository(ctx context.Context, projectID string) (*Repository, error) {
+	client, err := firestore.NewClient(ctx, projectID)
+	if err != nil {
+		log.Fatalf("Failed to create client: %v", err)
```
[nitpick] return the error instead?

GregoryAlbouy commented on firestore/repository.go:
```diff
@@ -0,0 +1,55 @@
+}
+
+func (r Repository) Create(ctx context.Context, data server.Report) (string, error) {
+	ref, _, err := r.client.Collection("reports").Add(ctx, data)
```
[totally ignorable suggestion] `"reports"` or even `r.client.Collection("reports")` could be referenced somewhere

GregoryAlbouy commented on http/handlers.go:
```diff
@@ -1,32 +1,35 @@
+	o := server.Output{}
+
+	err := gob.NewDecoder(r.Body).Decode(&o)
+	if err != nil && err != io.EOF {
```
```suggestion
	if err != nil {
```
(`Decode`'s `io.EOF` is not expected)

GregoryAlbouy commented on runner.go:
```diff
@@ -0,0 +1,67 @@
+	"time"
+)
+
+type Output struct {
```
`RunnerOutput`? `RunnerResult`? `BenchmarkResult`? `BenchmarkResult`!
But it could be `BenchmarkResult` as well.

moreirathomas replied:
Won't `gob` complain ? By extension: ins't the name of the struct and it's field part of `gob` encoding ?

GregoryAlbouy replied:
It won't, only the type itself matter!
Edit: the fields obviously but we don't change them? Not sure I got this question right!

moreirathomas replied:
I'm ok with either `RunResult` or `RunOutput`. No `Benckmark` because a benchmark is by definition a comparison of different implementation, which is not what the single run does. Hence `Run` is more appropriate to me.

Leaving `Benchmark` free for the future.

What do you prefer ?

And yes the actual fields would not change :)

GregoryAlbouy replied:
Let's go with `RunResult` then! Should we also rename `benchttp` to `runttp` meanwhile? 😛 
Edit: is it too _hair-pulled_ to note that `BenchmarkResult` could also be read as "single result of a benchmark"?

moreirathomas replied:
Arf f-it `Benchmark` is cool too and as a better readability:
- `benchttp.Benchmark.Report.Records`, `benchttp.Benchmark.Metadata.Config` vs
- `benchttp.RunResult.Report.Records`, `benchttp.RunResult.Metadata.Config`

Final call ?

GregoryAlbouy replied:
Haha, honestly both are fine to me! Maybe a little preference for the one that has `Result` in its name... But the `Benchmark` version reads better when accessing fields indeed.

On a side note, I didn't put much thinking into the structure when doing `Metadata.Config`, could as well be `Meta.Config`, or even flattened `Output.Report`, `Output.Config`. It's probably the time to discuss the final shape of that output (including `*http.Request` and `*http.Response` objects): Discord?

GregoryAlbouy commented on runner.go:
```diff
@@ -0,0 +1,67 @@
+	Time   time.Duration `json:"time" firestore:"time"`
+	Code   int           `json:"code" firestore:"code"`
+	Bytes  int           `json:"bytes" firestore:"bytes"`
+	Error  string        `json:"error,omitempty" firestore:"error,omitempty"`
+	Events []Event       `json:"events" firestore:"events"`
+}
+
+// Event is a stage of an outgoing HTTP request associated with a timestamp.
+type Event struct {
+	Name string        `json:"name" firestore:"name"`
+	Time time.Duration `json:"time" firestore:"time"`
+}
+
+type Config struct {
+	Request Request
+	Runner  Runner
+	Output  struct {
+		Out      []string
+		Silent   bool
+		Template string
+	}
+}
+
+type Request struct {
+	Method string
+	URL    *url.URL
+	Header http.Header
+	Body   struct {
+		Type    string
+		Content []byte
+	}
+}
+
+type Runner struct {
+	Requests       int
+	Concurrency    int
+	Interval       time.Duration
+	RequestTimeout time.Duration
+	GlobalTimeout  time.Duration
+}
```
I would declare these directly inside `Output` to reduce noise, as they only serve as a decoding structure for incoming requests.

moreirathomas replied:
The mega nesting it would produce is very noisy too and much less readable 😕
Would not exporting them be better ? I guess `gob` won't complain right ?

GregoryAlbouy replied:
I don't necessarily agree with the readability loss, as it also avoids having to navigate through the file to grasp the final structure, but anyway readability is not really a concern here because it's only a receiver structure for decoding the input request and store it as is.
Not exporting them is enough to me (and `gob` shouldn't complain about that as long as the fields themselves remain exported). However it might be a bit awkward to "export unexported types" (via `Output` fields).

GregoryAlbouy commented on ondisk/repository.go:
```diff
@@ -1,27 +1,32 @@
-package repository
+package ondisk
```
👋 🗑️

moreirathomas replied:
Do you suggest to completely delete `ondisk` and `kvstore` right now?

GregoryAlbouy replied:
Given we have no use for that I would 🤷‍♂️

GregoryAlbouy reviewed (commented):
Looks good 👌 

Some questions and nitpicks here and there.

GregoryAlbouy commented:
> By using `gob` between `runner` and `server`, the structs are _coupled across repos_, e.g. the fact that we need to change `runner` field types because Firestore does not support it (benchttp/runner#69). But `runner` should not really know, care or adapt to that.
Subject to discussion.

Well to me there's necessarily some sort of coupling as long as they communicate directly together, should it be `json` or `gob` encoded: if a field changes it must be done both sides — but it's true `gob`'s embarked typing adds more constraints. Let's be pragmatic: if it annoys us more than it serves us, we get rid of it.
I agree `runner` doesn't have to know about Firestore limitations: it's indeed the job of `server` to adapt its input to how it stores it. But again let's be pragmatic: if the simpler way is to update the runner, and that change does not introduce **explicit** coupling with an implementation detail of `server`, then it's better to do it rather than implement a silly adapter `server` side to e.g. change an `error` to `string`.
As long as we're not introducing logics/types that are specific to `server` implementation in `runner`, I wouldn't say `runner` depends on `server`, rather convenient input/output matching to avoid superfluous adapting complexity.

moreirathomas commented:
I totally agree with your point on struct coupling. At the moment it is _just_ definition that must match each other because both use the entire structure (one creates it, the other stores it as a whole).

I wanted to raise the attention around potential worse coupling in the future, because I was confronted to a slight friction when working on this feature in `server`.

This is fine and acceptable right now, but let's keep an eye on this !

moreirathomas committed:
misc: move http package to root

- move pkg/httplog inside new http package

moreirathomas committed:
feat: remove on disk store and repository

moreirathomas committed:
feat: update struct to reflect benchttp/runner

- define Benchmark struct as top level config
- redefine config related structs

moreirathomas committed:
feat: define firestore repository

- wrap official go lib inside custom firestore package
- custom firestore package implements server.Repository

moreirathomas committed:
feat: http.Server uses Repository

- pass env vars from main
- GOOGLE_APPLICATION_CREDENTIALS is read automatically

moreirathomas committed:
feat: call repository methods in server handlers

moreirathomas committed:
fix: mismatch Bencmark struct tags

moreirathomas committed:
misc: update handlers definition

c1-ra commented on benchttp.go:
```diff
@@ -0,0 +1,68 @@
+			} `firestore:"-"`
+		} `firestore:"config"`
+
+		FinishedAt time.Time `firestore:"finisedAt"`
```
An h is missing!

c1-ra commented on firestore/repository.go:
```diff
@@ -0,0 +1,76 @@
+	return repo, nil
+}
+
+// CloseClient calls the Close method of the interal client
```
"internal"

c1-ra commented on http/server.go:
```diff
@@ -33,19 +34,27 @@ func (s *Server) Start() error {
 	rw.Header().Set("Content-Type", "text/html; charset=utf-8")
 	rw.WriteHeader(200)
-	rw.Write([]byte("🖕")) //nolint
+	rw.Write([]byte("⚡")) //nolint
```
Ok for :zap:!

c1-ra reviewed (approved):
Nice! (I only corrected typos...)

moreirathomas committed:
misc: typos

moreirathomas commented on http/server.go:
```diff
@@ -33,19 +34,27 @@ func (s *Server) Start() error {
-	s.router.HandleFunc("/report", s.handleGetReport).Methods("GET")
-	s.router.HandleFunc("/report", s.handlePostReport).Methods("POST")
+
+	s.router.HandleFunc("/benchmark", s.handleRetrieveBenchmark).Methods("POST")
+
+	s.router.HandleFunc("/benchmark", s.handleCreateBenchmark).Methods("GET").
```
I realised that this change breaks the runner.
I'll keep it here because it makes more sense, but this opens the question of versioning our server endpoints and whether or not the runner must use env variables to provide the endpoint to call.

GregoryAlbouy replied:
- Versioning the endpoints is a perfectly suited solution for this, because we need the older endpoints to be always available for the runner
- We'll never use environment variables for the runner's target endpoint because they're user-defined and asking them to provide our new endpoint because we casually deleted the previous one doesn't sound like a great idea 😕

moreirathomas committed:
fix: swapped handlers

moreirathomas committed:
misc: reflect runner current endpoints expectations

moreirathomas commented:
There's some de-sync between runner and server with the state of this PR.

In order to not bloat the scope here, it will be addressed in another one.
